kitsune/tidings/events.py

```diff
@@ -121,29 +134,37 @@ def fire(self, exclude=None, delay=True):
         :arg exclude: If a saved user is passed in, that user will not be
           notified, though anonymous notifications having the same email
           address may still be sent. A sequence of users may also be passed in.
-
-        :arg delay: If True (default), the event is handled asynchronously with
-          Celery. This requires the pickle task serializer, which is no longer
-          the default starting in Celery 4.0. If False, the event is processed
-          immediately.
         """
-        if delay:
-            # Tasks don't receive the `self` arg implicitly.
-            self._fire_task.apply_async(
-                args=(self,), kwargs={"exclude": exclude}, serializer="pickle"
-            )
-        else:
-            self._fire_task(self, exclude=exclude)
-
-    @shared_task
-    def _fire_task(self, exclude=None):
-        """Build and send the emails as a celery task."""
         connection = mail.get_connection(fail_silently=True)
         # Warning: fail_silently swallows errors thrown by the generators, too.
         connection.open()
         for m in self._mails(self._users_watching(exclude=exclude)):
             connection.send_messages([m])
 
+    def serialize(self):
+        """
+        Serialize this event into a JSON-friendly dictionary.
+        """
+        result = dict(
+            event={"module": self.__class__.__module__, "class": self.__class__.__name__}
+        )
+        if instance := self.get_constructor_instance():
+            result.update(
+                instance={
+                    "module": instance.__class__.__module__,
+                    "class": instance.__class__.__name__,
+                    "id": instance.id,
+                }
+            )
+        return result
+
+    def get_constructor_instance(self):
+        """
+        Event subclasses constructed from a database model instance should override this
+        method and provide the database model instance used to construct themselves.
+        """
+        return None
+
```

Comment:
@tasos Based on our discussion earlier today, you might prefer that we get rid of `get_constructor_instance()` and make `serialize()` an abstract method that must be implemented in each subclass, so for example, each subclass would do something like this:

```python
def serialize(self):
    """
    Serialize this event into a JSON-friendly dictionary.
    """
    return {
        "event": {
            "module": "kitsune.wiki.events"
            "class": "EditDocumentEvent"
        },
        "instance": {
            "module": "kitsune.wiki.models",
            "class": "Revision",
            "id": self.revision.id
        }
    }
```

and we can still do that if you think that's easier to understand -- in fact I already have that option implemented on my laptop-- but just wanted to see what you thought of this approach first.

Reply:
I think this is a bit clearer on intentions :+1: